### ISSUE TITLE
Please include examples of the bad old ways and what to do about them instead, generally.

### ISSUE THREAD

**Issue by rhires** [dotnet/EntityFramework.Docs#3566]
I wish this information about the client side vs server side evaluation was somehow louder, because I got the exception for the first time without know why it happened, and finding this documentation was difficult. Stackoverflow doesn't have a good history of handling this issue, since it's relatively new. 

I wish there were more examples of what used to be "acceptable" but poor performance, and what is now totally blocked - as well as examples of what to do instead of the now totally blocked processes. 

I have a particular issue that uses the `.Where()` clause with an `.Include()`, and I can't find examples of what to do instead. I'm also sure that there are plenty of other places that an example would go a long way to helping people figure out what to do instead of the bad old ways :-) 

Thanks!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: fd35b912-445c-e801-7f70-de73839db80d
* Version Independent ID: e558a081-5397-1e3f-1d58-abea3f72cad6
* Content: [Client vs. Server Evaluation - EF Core](https://docs.microsoft.com/en-us/ef/core/querying/client-eval)
* Content Source: [entity-framework/core/querying/client-eval.md](https://github.com/dotnet/EntityFramework.Docs/blob/main/entity-framework/core/querying/client-eval.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @smitpatel
* Microsoft Alias: **avickers**

**Comment by roji** [comment]
@rhires prior to 3.0, EF Core used to automatically switch to client evaluation whenever something wasn't translatable to the server (SQL); all these cases are now blocked (except for top-level projections). Since this encompasses a very wide range of scenarios, it simply isn't practical (or particularly useful, I think) to document comprehensively. Assuming a user is aware of the change to disallow (most) client evaluation - which is mentioned in the release notes - it seems to me that the exception is explicit enough in saying the expression could not be translated, and additional docs wouldn't necessarily help.

Now, whether client evaluation is "acceptable but poor performance" is entirely dependent on your specific scenario and requirements. For example, a Where clause that cannot be translated previously caused the entire table to be pulled back for client-side evaluation; if the table happens to be small, that's fine, but if it's huge, this is likely to be a real perf-killer. So once again, I'm not sure what we could add in the documentation that would be helpful.

I'm also not sure what you're referring to with using Where and Include - AFAIK that should work just fine in most scenarios.

To summarize, it would good to understand very concretely what you're proposing we add to the docs etc.

**Comment by rhires** [comment]
Additionally, when I've looked for examples of what to do with a query like 

    ```
 var aspNetUsers = await _context
				.Users
                                .Where(m => m.Id = userId)
				.Include(u => u.UserRoles)
				.Select(m => new AppUserViewModel
				{
					StaffId = m.StaffId,
					Id = id
     });

```
They simply take out the Where() clause, do the projection, then do the filter...which happens client side. Everyone says that this can cause performance problems when datasets are large, but no one seems to have a solution for how to do server side filtering, especially for large datasets, to avoid the performance issue. I know it can't be done in every case, but is there a solution for this?

**Comment by roji** [comment]
> They simply take out the Where() clause, do the projection, then do the filter...which happens client side. Everyone says that this can cause performance problems when datasets are large, but no one seems to have a solution for how to do server side filtering, especially for large datasets, to avoid the performance issue. I know it can't be done in every case, but is there a solution for this? 

@rhires can you point to where the above query is said to be problematic? Everything in the above query should get translated to SQL and execute efficiently at the server, with only the StaffId column getting sent to the client for rows with the matching Id. Have you executed the query and seen something else in the generated SQL?

It's also not clear why you're including UserRoles, as it isn't used later.

**Comment by rhires** [comment]
Ok - while I wasn't trying to get support for this query, and it was an example of a "bad" query that would generate an exception, I re-ran it the way I thought it was giving me errors, and it no longer is. So I'm quite puzzled by that. I also just posted an abbreviated example. I was getting UserRoles to associate them with the users in my admin interface on the UI side, it's just not shown in my example. I'll remember to post a more complete example next time. Sorry about that.

But my main request... It would be nice to have examples of the previously allowed but now disallowed situations, and what the "replacement" query might be. The text in the doc is well written, and explains the rationale pretty well about why the change, but it doesn't say what to do instead now that the change is in place. 

I realize that not all situations can be documented, but one of the first things would be to have a specific page for the error "The LINQ expression could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation"...and then some explanations of what might be wrong with the query. It might also be useful to change the message to be closer to what the error really is: some kind of evaluation that EF is doing not at the top level, but a level or two below the top level which it can't send down to the server. 

Here's also [an different example](https://stackoverflow.com/questions/60525592/the-linq-expression-could-not-be-translated-either-rewrite-the-query-in-a-form) from stackoverflow to maybe better illustrate what I'm talking about. The exception has the user scratching their head at what that even means, and then the person answering taking a guess about how to fix the issue. 

So there must be a small list of these kind of "second tier" evaluations/projections that are common that would generate an exception like the one on SO as shown above, that would go on a page similar to the [page](https://www.thecodebuzz.com/query-linq-expression-could-not-be-translated/) by a third party that better explains what the error means, and what could generate it.

**Comment by ajcvickers** [comment]
@rhires Are you asking how to use `AsEnumerable` to cause client evaluation? Or are you asking how to rewrite queries so they don't need client evaluation?

**Comment by rhires** [comment]
No, not asking how to do that explicitly. I'm looking more for the "rewrite queries so they don't need client evaluation", but in the form of examples - examples of what used to work pre-.net 3.1 (3.0?), and now won't work because of the new logic. And then how they would now look with the new logic in place. 

I suspect that there are common patterns that devs use that will throw the exception, that could be documented as the "old way", and then shown with the "new way". 

So we could have a page with the name of the exception in the title, explaining what "cannot be translated" means - that LINQ cannot convert the LINQ as written into SQL and passed down to the server for evaluation. Then we have an example (as on the third party page linked above), to show how throwing a method into the middle of a LINQ statement causes a problem, since it can't be converted to SQL (but its result can, so make it a stand-alone statement). 

Then we could have another example of something that used to be allowed in say, 2.1, but isn't any longer. And then a few more after that. Two columns: "old way" | "new way". 

Similarly, but not the same, there are a few pages in the docs somewhere that show quite a few LINQ simple examples for the different operators, but these examples don't have an old vs new. 

I'm sorry I'm having such a hard time communicating this out to you. I appreciate you trying to understand what I'm talking about.

**Comment by roji** [comment]
@rhires I think I understand what you're asking. However, given that it would be quite a big effort to write this, what do you see are the exact advantages of this, given that users can simply run their queries and see them fail? Our "could not translate" exception messages are probably not perfect, but they do point to the specific part of the query which cannot be translated. I agree this doesn't tell users how to fix the issue, but from our experience, users generally figure it out (e.g. they execute some local .NET function inside a Where clause, and understand that this cannot be translated).

So I'm not saying we're perfect here, but overall I'd rather we invested in tailoring exception translation failure messages rather than attempting to cover everything via docs.

**Comment by rhires** [comment]
I think that’s probably the better option: updating the exception translation message to be a little less cryptic, such as you mention below. Basically, I couldn’t immediately figure out what the exception message meant, since I don’t think of LINQ as something that gets “translated”. 

Also, I don’t believe I saw which specific part of the query gets pointed to that is causing the issue in the exception or the stack trace. I probably just missed it, because I didn’t understand the “headline" of the exception. 

Thank you - I guess that’s all I have. Cheers!

**Comment by roji** [comment]
> since I don’t think of LINQ as something that gets “translated”.

Yeah, that's basically the main role of EF Core when it comes to queries - translate LINQ queries to SQL.

In any case, thanks for engaging and don't hesitate to reach out with other issues.
